WooCommerce/Classes/ViewModels/MainTabViewModel.swift

```diff
@@ -42,39 +55,68 @@ final class MainTabViewModel {
     ///
     func startObservingOrdersCount() {
         observeBadgeRefreshNotifications()
+        updateBadgeFromCache()
         requestBadgeCount()
     }
 }
 
 
 private extension MainTabViewModel {
+
+    /// Construct `ResultsController` with `siteID`
+    ///
+    func updateStatusResultsController(siteID: Int64) -> ResultsController<StorageOrderStatus> {
+        let predicate = NSPredicate(format: "siteID == %lld AND slug == %@", siteID, OrderStatusEnum.processing.rawValue)
+        return ResultsController<StorageOrderStatus>(storageManager: ServiceLocator.storageManager, matching: predicate, sortedBy: [])
+    }
+
+    /// Connect hooks on `ResultsController` and query cached data
+    ///
+    func configureResultsController() {
+        statusResultsController?.onDidChangeObject = { [weak self] (updatedOrdersStatus, _, _, _) in
+            self?.processBadgeCount(updatedOrdersStatus)
+        }
+
+        try? statusResultsController?.performFetch()
+        updateBadgeFromCache()
+    }
+
+    /// Get last known data from cache (if exists) and draw it on a badge
+    ///
+    func updateBadgeFromCache() {
+        let initialCachedOrderStatus = statusResultsController?.fetchedObjects.first
+        processBadgeCount(initialCachedOrderStatus)
+    }
+
+    /// Trigger network action to update underlying cache. Badge redraw will be triggered by `statusResultsController`
+    ///
     @objc func requestBadgeCount() {
         guard let siteID = storesManager.sessionManager.defaultStoreID else {
             DDLogError("# Error: Cannot fetch order count")
             return
         }
 
-        let action = OrderAction.countProcessingOrders(siteID: siteID) { [weak self] orderCount, error in
-            if error != nil {
-                return
+        let action = OrderStatusAction.retrieveOrderStatuses(siteID: siteID) { result in
+            if case let .failure(error) = result {
+                DDLogError("⛔️ Could not successfully fetch order statuses for siteID \(siteID): \(error)")
             }
-
-            self?.processBadgeCount(orderCount)
```

Comment:
Now we don't process immediate callback, we listen for Core Data updates to get any related changes.
This network request will be throttled later as a part of networking improvements.